### DISCUSSION TITLE
Why are UsingStatements in Constrained LanguageMode allowed?

### DISCUSSION THREAD

**Discussion by Herrmel** [PowerShell/PowerShell#16100] [Q&A] [answered]
I was wondering if it is intended to allow UsingStatements in the Constrained LanguageModes?
UsingStaments allow to import any Assembly as simple as `using assembly <path>`. Since stuff like `Add-Type -Path <path>` is forbidden in this LanguageMode I feel like the UsingStatement should be as well, maybe even in the Restricted LanguageMode?

**Comment by vexx32** [comment]
That's probably a question for @PaulHigin, but it would probably make sense to disable `using assembly` in constrained language mode.

**Comment by PaulHigin** [comment, marked as answer]
My initial take is that this restriction is not needed.  Even when assemblies are loaded into the PowerShell process, under `Constrained Language mode` and via the `using` statement, the assembly types are not accessible either through Add-Type or public API.  

A lot of assemblies are loaded into a PowerShell process by various means, but Constrained Language mode prevents access to types exposed by those assemblies.

**Reply by Herrmel** [reply]
Thanks for the quick response. Public static Properties and Fields can be accessed though, but I guess this should not be too dangerous as long as the user is not be able to add his own assembly to the filesystem somehow.
I built the Powershell into my Programm to enable a lot more possibilities in the Consoleinterface wich works really great. I'm now trying to prevent users from importing assemblies and Modules that I did not explicitly allow but I can't disable the UsingStatements. I guess I'll have to analyze the AST from userinput beforehand and search there for UsingStatements to disable them. Does not feel like the cleanest Way but I guess it will do.
